numpy/core/src/multiarray/compiled_base.dispatch.c

```diff
@@ -0,0 +1,99 @@
+/**
+ * @targets $maxopt baseline
+ * SSE2 AVX2
+ * NEON ASIMDDP
+ */
+#include "compiled_base.h"
+
+/*
+ * This function packs boolean values in the input array into the bits of a
+ * byte array. Truth values are determined as usual: 0 is false, everything
+ * else is true.
+ */
+NPY_NO_EXPORT void NPY_CPU_DISPATCH_CURFX(compiled_base_pack_inner)
+(const char *inptr, npy_intp element_size,  npy_intp n_in, npy_intp in_stride, char *outptr, npy_intp n_out, npy_intp out_stride, char order)
+{
+    /*
+     * Loop through the elements of inptr.
+     * Determine whether or not it is nonzero.
+     *  Yes: set corresponding bit (and adjust build value)
+     *  No:  move on
+     * Every 8th value, set the value of build and increment the outptr
+     */
+    npy_intp index = 0;
+    int remain = n_in % 8;              /* uneven bits */
+
+#if defined(NPY_HAVE_SSE) || defined(NPY_HAVE_AVX2) || defined(NPY_HAVE_NEON)
+    if (in_stride == 1 && element_size == 1 && n_out > 2) {
+        npyv_u64 zero = npyv_zero_u64();
+        /* don't handle non-full 8-byte remainder */
+        npy_intp vn_out = n_out - (remain ? 1 : 0);
+        const int vstep = npyv_nlanes_u64;
+        vn_out -= (vn_out & (vstep - 1));
+        for (index = 0; index < vn_out; index += vstep) {
+            // Maximum paraller abillity: handle four 64bits at one time
+            npy_uint64 a[4];
+            for (int i = 0; i < vstep; i++) {
+                a[i] = *(npy_uint64*)(inptr + 8 * i);
+                if (order == 'b') {
+                    a[i] = npy_bswap8(a[i]);
+                }
+            }
+            /* note x86 can load unaligned */
+            npyv_u64 v;
+            if (vstep == 4) {
+                v = npyv_set_u64(a[0], a[1], a[2], a[3]);
+            } else {
+                v = npyv_set_u64(a[0], a[1]);
+            }
+            /* false -> 0x00 and true -> 0xFF (there is no cmpneq) */
+            v = npyv_reinterpret_u8_u64(npyv_cmpeq_u8(npyv_reinterpret_u8_u64(v), npyv_reinterpret_u8_u64(zero)));
+            v = npyv_reinterpret_u8_u64(npyv_cmpeq_u8(npyv_reinterpret_u8_u64(v), npyv_reinterpret_u8_u64(zero)));
+            /* extract msb of 16 bytes and pack it into 16 bit */
+            unsigned int r = npyv_movemask_u8(npyv_reinterpret_u8_u64(v));
```

Comment:
```suggestion
            npyv_b8 bmask = npyv_cmpneq_u8(a, v_zero);
            unsigned int r = npyv_movemask_b8(bmask);
```
1. use `npyv_cmpneq_u8`(not equal) instead of two `npyv_cmpeq_u8`
2. NPYV has special vector types for boolean vectors to add compatibility with AVX512 and VSX(Altivec bool hint):
    1. **npyv_b8**   returned from (u8 and s8)  compersion operations
    2. **npyv_b16** returned from (u16 and s16)  // //
    3. **npyv_b32** returned from (u32, s32 and f32) // //
    4. **npyv_b64** returned from (u64, s64 and f64) // //

You can count on `npyv_cvt_*_*` if you want to convert between integer and boolean vectors.
However, in this case, you just need to rename suffix of `npyv_movemask_u8` to be `b8` and
add support for AVX512.

Just for the record, NPYV doesn't have support for type safety instead it counts on the SIMD extensions itself and the `_SIMD` module(#16782).